ucp/public_api.py

```diff
@@ -261,18 +293,59 @@ def abort(self):
         Notice, this functions doesn't signal the connected peer to close.
         To do that, use `Endpoint.close()`
         """
-        self._ep.abort()
+        if self._closed:
+            return
+        self._closed = True
+        logging.debug("Endpoint.abort(): %s" % hex(self.uid))
+
+        for msg in self.pending_msg_list:
+            if "future" in msg and not msg["future"].done():
+                logging.debug("Future cancelling: %s" % msg["log"])
+                self._worker.request_cancel(msg["ucp_request"])
+
+        self._ep.close(self._worker)
+        self._ep = None
+        self._ctx = None
+        self._worker = None
 
     async def close(self):
         """Close the endpoint cleanly.
         This will attempt to flush outgoing buffers before actually
         closing the underlying UCX endpoint.
         """
-        await self._ep.close()
+        if self._closed:
+            return
+        try:
+            # Send a shutdown message to the peer
+            from .application_context import CtrlMsg
+
+            msg = CtrlMsg.serialize(opcode=1, close_after_n_recv=self._send_count)
+            log = "[Send shutdown] ep: %s, tag: %s, close_after_n_recv: %d" % (
+                hex(self.uid),
+                hex(self._ctrl_tag_send),
+                self._send_count,
+            )
+            logging.debug(log)
+            self.pending_msg_list.append({"log": log})
+            try:
+                await send_recv.tag_send(
+                    self._ep,
+                    msg,
+                    len(msg),
+                    self._ctrl_tag_send,
+                    pending_msg=self.pending_msg_list[-1],
+                )
+            except exceptions.UCXError:
+                pass  # The peer might already be shutting down
+            # Give all current outstanding send() calls a chance to return
+            self._ctx.progress()
+            await asyncio.sleep(0)
```

Comment:
Why is `sleep(0)` needed here?

Reply:
To give other co-routines a chance to run before calling `self.abort()`, which cancel pending messages.

(Updated the code comment )

Reply:
Is `0` enough time or should we be choosing some small number here?